### PR TITLE
NFSD Access Overhaul

### PR DESCRIPTION
**Pull request by neuPanda**
## About the PR
NFSD Access Overhaul
Fixed ship console refresh on id insert or removal
Added Access to VesselPrototype
Added Filtration to Ship List based off of Access
Added Bailiff and Sergeant Access levels
Added Access levels to all NFSD ship
Added appropriate access groups to NFSD roles

## Why / Balance
In Preperation for incoming NFSD changes
## How to test

Go to NFSD ship yard
using each of the NFSD ID cards swap back and fort to observe available ships
ensure that non NFSD has no access
buy an empress and then test again with the empress ships
verify the new permissions for each of the NFSD roles at the ID console

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- fix: ship console refresh on id insert or removal
- add: Filtration to Ship List based off of Access
- add: Bailiff and Sergeant Access levels

### CODE REVIEW

**Comment by dvir001** [comment]
Add ``// Frontier`` to every single added line to code block

File: Content.Shared/Access/Components/IdCardConsoleComponent.cs [hunk 1]
```diff
@@ -3,8 +3,6 @@
 using Robust.Shared.GameStates;
 using Robust.Shared.Prototypes;
 using Robust.Shared.Serialization;
-using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype.List;
-using Robust.Shared.Prototypes;
```
Comment (dvir001) [inline Content.Shared/Access/Components/IdCardConsoleComponent.cs:7]: Dont remove this even if its unused if its from upsteam.

File: Content.Shared/Shipyard/Prototypes/VesselPrototype.cs [hunk 2]
```diff
@@ -39,6 +40,12 @@ public sealed class VesselPrototype : IPrototype
     [DataField("group")]
     public string Group = string.Empty;
 
+    /// <summary>
+    ///     The access of the product. (e.g. Command, Mail, Bailiff, etc.)
```
Comment (dvir001) [inline Content.Shared/Shipyard/Prototypes/VesselPrototype.cs:44]: Frontier mention

File: Content.Shared/Shipyard/Prototypes/VesselPrototype.cs [hunk 3]
```diff
@@ -51,4 +58,5 @@ public sealed class VesselPrototype : IPrototype
     /// </summary>
     [DataField("shuttlePath", required: true)]
     public ResPath ShuttlePath = default!;
+
```
Comment (dvir001) [inline Content.Shared/Shipyard/Prototypes/VesselPrototype.cs:61]: No need extra space here.

File: Resources/Prototypes/_NF/Roles/Jobs/Nfsd/bailiff.yml [hunk 4]
```diff
@@ -25,6 +25,9 @@
   - Mercenary # Frontier
   - Captain # Frontier
   - Armory
+  accessGroups: # Frontier
+  - GeneralNfsdAccess # Frontier
+  - ExecutiveNfsdAccess # Frontier
```
Comment (dvir001) [inline Resources/Prototypes/_NF/Roles/Jobs/Nfsd/bailiff.yml:30]: Did you test
``access``
``accessGroups``

Working on the same time? You dont need them to exists at the same time.

File: Resources/Prototypes/_NF/Roles/Jobs/Nfsd/sheriff.yml [hunk 5]
```diff
@@ -26,6 +26,8 @@
   canBeAntag: false
   accessGroups:
   - AllAccess
+  - GeneralNfsdAccess # Frontier
+  - CommandNfsdAccess # Frontier
```
Comment (dvir001) [inline Resources/Prototypes/_NF/Roles/Jobs/Nfsd/sheriff.yml:30]: ``AllAccess``

**Review by dvir001 — CHANGES_REQUESTED** [review]

File: Content.Shared/Access/Components/IdCardConsoleComponent.cs [hunk 6]
```diff
@@ -47,7 +45,8 @@ public WriteToTargetIdMessage(string fullName, string jobTitle, List<ProtoId<Acc
     public List<ProtoId<AccessLevelPrototype>> AccessLevels = new()
     {
         "Armory",
-        //"Atmospherics",
+        "Atmospherics",
```
[the remark this answers is not in the document]
Comment (neuPanda) [reply Content.Shared/Access/Components/IdCardConsoleComponent.cs:48]: its causing an issue where SR can corrupt their access and lose access to engineering on station

File: Content.Shared/Access/Components/IdCardConsoleComponent.cs [hunk 7]
```diff
@@ -59,7 +58,7 @@ public WriteToTargetIdMessage(string fullName, string jobTitle, List<ProtoId<Acc
         //"ChiefMedicalOfficer",
         "Command",
         //"Cryogenics",
-        //"Engineering",
+        "Engineering",
```
[the remark this answers is not in the document]
Comment (neuPanda) [reply Content.Shared/Access/Components/IdCardConsoleComponent.cs:61]: its causing an issue where SR can corrupt their access and lose access to engineering on station

File: Content.Shared/Shipyard/Prototypes/VesselPrototype.cs (hunk 2 above)
Comment (dvir001) [reply Content.Shared/Shipyard/Prototypes/VesselPrototype.cs:44]: Looking into this this all file should be moved under _NF but that a different issue.

File: Content.Shared/Access/Components/IdCardConsoleComponent.cs (hunk 1 above)
Comment (neuPanda) [reply Content.Shared/Access/Components/IdCardConsoleComponent.cs:7]: this must have been done by re-sharper

File: Resources/Prototypes/_NF/Roles/Jobs/Nfsd/bailiff.yml (hunk 4 above)
Comment (neuPanda) [reply Resources/Prototypes/_NF/Roles/Jobs/Nfsd/bailiff.yml:30]: yes i tested

File: Content.Shared/Access/Components/IdCardConsoleComponent.cs (hunk 1 above)
Comment (neuPanda) [reply Content.Shared/Access/Components/IdCardConsoleComponent.cs]: resolved

File: Resources/Prototypes/_NF/Roles/Jobs/Nfsd/sheriff.yml (hunk 5 above)
Comment (neuPanda) [reply Resources/Prototypes/_NF/Roles/Jobs/Nfsd/sheriff.yml:30]: was not able to find AA and figured it was somewhere on the upstream code, and i am trying to keep frontier changes to frontier.  

In addition, not 100% sold on the concept of AA
IMO Sheriff and SR should be able to have their own private rooms or ships.  
AA rejects that concept.  if everybody was still on frontier station them yah, maybe.  but now they have 2 separate stations.  and with more potential command roles coming, i more strongly support the idea of separation of access

File: Content.Shared/Access/Components/IdCardConsoleComponent.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (dvir001) [reply Content.Shared/Access/Components/IdCardConsoleComponent.cs:63]: The fix is to remove it from frontier station not to add it to ID computer.

File: Content.Shared/Access/Components/IdCardConsoleComponent.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dvir001) [reply Content.Shared/Access/Components/IdCardConsoleComponent.cs:50]: The fix is to remove it from frontier station not to add it to ID computer.

File: Content.Shared/Access/Components/IdCardConsoleComponent.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (dvir001) [reply Content.Shared/Access/Components/IdCardConsoleComponent.cs:63]: https://github.com/new-frontiers-14/frontier-station-14/pull/1461

File: Content.Shared/Access/Components/IdCardConsoleComponent.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dvir001) [reply Content.Shared/Access/Components/IdCardConsoleComponent.cs:50]: https://github.com/new-frontiers-14/frontier-station-14/pull/1461

[the remark this answers is not in the document]
Comment (dvir001) [reply Content.Shared/Access/Components/IdCardConsoleComponent.cs:50]: Fixed, please un comment it again

File: Content.Shared/Access/Components/IdCardConsoleComponent.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (dvir001) [reply Content.Shared/Access/Components/IdCardConsoleComponent.cs:63]: Fixed, please un comment it again

**Review by dvir001 — DISMISSED** [review]

**Comment by MagnusCrowe** [comment]
image
SR has every access except Bailiff and Sergeant?! Sheriff has more access than the Station Representative?!
Yet, SR still has Detective access?  Please explain your reasoning.

The ships Sheriff can access:
- All from the NFSD shipyard
- All from the Empress shipyard
The ships the Bailiff can access:
- All but detective ships from the NFSD shipyard?!  Explain.
- All from the Empress shipyard
The Sergeant can access:
- Prowler and Templar from the NFSD shipyard...no Hospitalier?
- All from the Empress shipyard
The Deputy can access:
- Nothing from the NFSD shipyard?!  Not even a Templar? Explain.
- All from the Empress shipyard
The Brigmedic can access:
- Nothing from the NFSD shipyard?!  Not even a Hospitalier? Explain.
- All from the Empress shipyard

I don't understand the reasoning for these distinctions and lack of distinctions.  What is the end goal here?

The way I see NFSD ships, there are 3 categories (don't read too far into the category names, don't "erm, acktully" me with naval technicalities please):
- Flagships [Empress, Wasp] (almost always 1 per shift)
- Cruisers [Marauder, Prowler, Opportunity hehe] (usually 1 or 2 per shift)
- Support [Templar, Hospitalier, Interceptor, Inquisitor, Broadhead, Cleric, Fighter, Rogue] (many per shift)
There are categories of NFSD personnel too:
- Sheriff
- Officers [Bailiff, Sergeant]
- Squad [Deputies, Cadets, Brigmedics, Detectives, Prison Guards]

I would reorganize the access as such:
- Sheriff [Flagships, Cruisers, Support]
- Bailiff [Cruisers, Support]
- Sergeant [Marauder, Prowler, Support]
- Deputy [Templar, Fighter, Rogue]
- Brigmedic [Hospitalier, Cleric]
- Detective [Interceptor, Inquisitor, Broadhead]
- Prison Guard [None, what if a prisoner got their ID, get ships from bailiff]
- Cadet [None, you shouldn't be operating your own ship before graduating the academy]

With these access levels every role has access to the ships they would typically pilot.  If the Sheriff orders any of these jobs to procure a specific ship related to their job, they have access to do that.  Each commander has the ability to purchase ships on behalf of the jobs that would be commanded by them. Let me know what you thing and what your reasoning was behind the choices you made.

**Comment by MagnusCrowe** [comment]
On a more technical note, merge in the latest changes and set up the broadhead with access levels.

**Comment by Cheackraze** [comment]
I agree with the suggested access changes actually, especially if we are going to be trying to steer the brigmedic away from officer mentality stuff or regular deputy stuff

**Comment by MagnusCrowe** [comment]
image
Bailiff not intended to have detective access? Fine with me just seemed strange and not sure if it was intended...
All of my intention questions could be cleared up by stating what your intended changes are in the PR's About section.

**Comment by neuPanda** [comment]
>Bailiff not intended to have detective access? Fine with me just seemed strange and not sure if it was intended... All of my intention questions could be cleared up by stating what your intended changes are in the PR's About section.

correct, the detective works directly for the sheriff not the bailiff.
the detective needs to be able to investigate not only external issues but internal ones as well and can not have the bailiff interfering with their work

**Comment by neuPanda** [comment]
_points at github_
I will win you little shit stop adding merge conflicts that were not there moments ago

**Comment by neuPanda** [comment]
_glairs at github_

**Comment by neuPanda** [comment]
_pulls out the throngler_

**Review by whatston3 — CHANGES_REQUESTED** [review]
Throwing my hat in on this one.

I've taken a look through the code and come up with a set of suggestions on [whatston3:NFSD-access-expansion-tests-and-suggestions](https://github.com/whatston3/frontier-station-14/tree/NFSD-access-expansion-tests-and-suggestions)

There's a diff without the merge of the current head of master [up here](https://github.com/new-frontiers-14/frontier-station-14/compare/4c6e9ebb3a936af6d23d244e7aeef3f4def25230...whatston3:frontier-station-14:NFSD-access-expansion-tests-and-suggestions?expand=1)

In brief:
* Removed LINQ from the shipyard console code, replaced it with iterators and if statements.
* Added checks for the ship's access against a role's groups as well as their tags - the sheriff can now buy a ship despite not having any explicit access tags.
* Removed the officer/executive/sheriff security groups: too small to warrant separate groups at the moment to my mind, easier to just state outright which jobs have which accesses there.  Base NFSD one's great, cadet's marginal.
    * If the idea was to have all of the definitions in one file, I suppose that's fair, but you'd still need to know both who is in what group and what those accesses are.  Not hugely convinced of the value.
* Removed redundant accesses and groups from Sheriff, Detective
* Updated to the current head of master (mostly for testing)

In my (admittedly brief) testing so far, this seems to behave well.

**Review by whatston3 — APPROVED** [review]
Happy with the state of this.  Good feature.

**Pull request merged**